// angularfire2/__ivy_ngcc__/database-deprecated/es2015/database.js
import { Inject, Injectable, Optional, NgZone } from '@angular/core';
import { FirebaseListFactory } from './firebase_list_factory';
import { FirebaseObjectFactory } from './firebase_object_factory';
import * as utils from './utils';
import { FirebaseOptionsToken, FirebaseNameOrConfigToken, RealtimeDatabaseURL, _firebaseAppFactory } from 'angularfire2';
import * as ɵngcc0 from '@angular/core';
export class AngularFireDatabase {
    constructor(options, nameOrConfig, databaseURL, zone) {
        this.database = zone.runOutsideAngular(() => {
            const app = _firebaseAppFactory(options, nameOrConfig);
            return app.database(databaseURL || undefined);
        });
    }
    list(pathOrRef, opts) {
        const ref = utils.getRef(this.database, pathOrRef);
        return FirebaseListFactory(ref, opts);
    }
    object(pathOrRef, opts) {
        const ref = utils.getRef(this.database, pathOrRef);
        return FirebaseObjectFactory(ref, opts);
    }
}
AngularFireDatabase.ɵfac = function AngularFireDatabase_Factory(t) { return new (t || AngularFireDatabase)(ɵngcc0.ɵɵinject(FirebaseOptionsToken), ɵngcc0.ɵɵinject(FirebaseNameOrConfigToken), ɵngcc0.ɵɵinject(RealtimeDatabaseURL, 8), ɵngcc0.ɵɵinject(ɵngcc0.NgZone)); };
AngularFireDatabase.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: AngularFireDatabase, factory: AngularFireDatabase.ɵfac });
AngularFireDatabase.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [FirebaseOptionsToken,] },] },
    { type: undefined, decorators: [{ type: Inject, args: [FirebaseNameOrConfigToken,] },] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [RealtimeDatabaseURL,] },] },
    { type: NgZone, },
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularFireDatabase, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [FirebaseOptionsToken]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [FirebaseNameOrConfigToken]
            }] }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [RealtimeDatabaseURL]
            }] }, { type: ɵngcc0.NgZone }]; }, null); })();
export { RealtimeDatabaseURL };
